noir-projects/aztec-nr/aztec/src/state_vars/shared_mutable/scheduled_value_change.nr

```diff
@@ -0,0 +1,279 @@
+            //           block                        block horizon      block of change
+            //        =====N=================================H=================|=========>
+            //              ^                                 ^
+            //              |                                 |   
+            //              -----------------------------------
+            //                           delay
+            //
+            //   b) It could be fewer than `delay` blocks away from the historical block number, in which case it would
+            //      become the limiting factor for the time horizon, which would be the block right before the block of
+            //      change (since by definition the value changes at the block of change).
+            //
+            //           historical                         block horizon
+            //             block     block of change       if not scheduled
+            //        =======N=============|===================H=================>
+            //                ^           ^                    ^
+            //                |     actual horizon             |   
+            //                ----------------------------------
+            //                             delay
+            //      
+            // Note that the current implementation does not allow the caller to set the block of change to an arbitrary
+            // value, and therefore scenario a) is not currently possible. However implementing #5501 would allow for 
+            // this to happen.
+
+            // Because historical_block_number < self.block_of_change, then block_of_change > 0 and we can safely
+            // subtract 1.
+            min(self.block_of_change - 1, historical_block_number + DELAY)
+        }
+    }
+
+    pub fn schedule_change(&mut self, new_value: T, current_block_number: u32) {
+        self.pre = self.get_current_at(current_block_number);
+        self.post = new_value;
+        // TODO: make this configurable 
+        // https://github.com/AztecProtocol/aztec-packages/issues/5501
+        self.block_of_change = current_block_number + DELAY;
+    }
+}
+
+impl<T, DELAY> Serialize<3> for ScheduledValueChange<T, DELAY> {
+    fn serialize(self) -> [Field; 3] where T: ToField {
+        [self.pre.to_field(), self.post.to_field(), self.block_of_change.to_field()]
+    }
+}
+
+impl<T, DELAY> Deserialize<3> for ScheduledValueChange<T, DELAY> {
+  fn deserialize(input: [Field; 3]) -> Self  where T: FromField {
+    Self { 
+        pre: FromField::from_field(input[0]), 
+        post: FromField::from_field(input[1]), 
+        block_of_change: FromField::from_field(input[2]), 
+        _dummy: [0; DELAY] 
+    }
+  }
+}
+
+fn min<T>(lhs: T, rhs: T) -> T where T: Ord {
+    if lhs < rhs { lhs } else { rhs }
+}
+
+#[test]
+fn test_min() {
+    assert(min(3, 5) == 3);
+    assert(min(5, 3) == 3);
+    assert(min(3, 3) == 3);
+}
+
+mod test {
```

Comment:
Would likely be nice to test a case with larger structs for the pre and post to ensure working as expected, looks mainly to be single fields here? 

It might be too disgusting to get working before we can do `N*2+1` as the size of the serialization though.

Reply:
Yeah `ScheduledValueChange` by itself would be fine since it doesn't use `T` for anything, the problem will be accesing storage due to incomplete support for numeric generics in Noir, as mentioned in https://github.com/AztecProtocol/aztec-packages/issues/5492.